client/txn.go

```diff
@@ -327,7 +336,8 @@ func (txn *Txn) exec(retryable func(txn *Txn) error) (err error) {
 		}
 		break
 	}
-	if err != nil && txn.haveTxnWrite {
+	if err != nil && txn.txn.Writing {
+		txn.haveEndTxn = false // if we sent one, it didn't succeed
```

Comment:
If the retry loop gave up, and we have a transaction running, then even if we tried to run `EndTransaction`, it must have failed (or was never run; after all, it's the last call to be executed). So we pretend we never sent one (this is necessary since otherwise that double-commit check would complain). I'll add that to the comment.